### DISCUSSION TITLE
Namespace of imported servce template

### DISCUSSION THREAD

**Discussion by foliveira3** [tliron/go-puccini#78] [Q&A] [answered]
### I have an ETSI SOL001 VNF Descriptor which imports the ETSI SOL001 vnfd types and subsequently  imports a file with some custom types.  The custom types are derived from the ETSI SOL001 types.  Puccini raises errors that the derived types are not defined.

> vnfd.yaml:
> tosca_definitions_version: tosca_simple_yaml_1_3
> imports:
>     - https://forge.etsi.org/rep/nfv/SOL001/raw/v3.3.1/etsi_nfv_sol001_vnfd_types.yaml
>     - custom_types.yaml # custom data types

-------------

> custom_types.yaml:
> tosca_definitions_version: tosca_simple_yaml_1_3
> node_types:
> 
>   custom.nodes.nfv.Vdu.Compute:
>     derived_from: tosca.nodes.nfv.Vdu.Compute
>     properties:
>       name_pattern:
>         type: string
>         description: Name pattern definition for this compute node_types
>         required: false

-----------------------------------------------
This yields an error of "reference to unknown node type: tosca.nodes.nfv.Vdu.Compute"

I was able to remove this error by adding an import of the ETSI definitions to the custom types.yaml file:

> Updated custom_types.yaml:
> tosca_definitions_version: tosca_simple_yaml_1_3
> _imports:
>     - https://forge.etsi.org/rep/nfv/SOL001/raw/v3.3.1/etsi_nfv_sol001_vnfd_types.yaml_
> 
> node_types:
> 
>   custom.nodes.nfv.Vdu.Compute:
>     derived_from: tosca.nodes.nfv.Vdu.Compute
>     properties:
>       name_pattern:
>         type: string
>         description: Name pattern definition for this compute node_types
>         required: false
> 
------------------------
Is this correct behavior or should the ETSI definitions imported in the parent "vnfd" be visible to the child "custom" definitions?

**Comment by tliron** [comment, marked as answer]
This is intentional behavior.

Every imported file (called a "unit" internally) in Puccini is its own universe. This is necessary in order to maintain proper support of TOSCA namespaces. An imported unit does not know in advance the document into which it may imported and must be fully parsable on its own. Thus, if custom_types.yaml refers to certain types then it *must* import them explicitly. It cannot reach them from the document into which it is imported.

Note that in terms of performance Puccini is very efficient and will internally not parse the same imported unit more than once (as long as its referred to be the same URL or path).

By the way, you can dump the import tree with the following command:
```
puccini-tosca parse -d 1 ...
```

**Reply by foliveira3** [reply]
Thanks
